src/hotspot/share/gc/shenandoah/heuristics/shenandoahReactiveHeuristics.cpp

```diff
@@ -0,0 +1,308 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+
+#include "gc/shenandoah/heuristics/shenandoahReactiveHeuristics.hpp"
+#include "gc/shenandoah/shenandoahCollectionSet.hpp"
+#include "gc/shenandoah/shenandoahFreeSet.hpp"
+#include "gc/shenandoah/shenandoahHeap.inline.hpp"
+#include "logging/log.hpp"
+#include "logging/logTag.hpp"
+#include "utilities/quickSort.hpp"
+
+// These constants are used to adjust the margin of error for the moving
+// average of the allocation rate and cycle time. The units are standard
+// deviations.
+const double ShenandoahReactiveHeuristics::FULL_PENALTY_SD = 0.2;
+const double ShenandoahReactiveHeuristics::DEGENERATE_PENALTY_SD = 0.1;
+
+// These are used to decide if we want to make any adjustments at all
+// at the end of a successful concurrent cycle.
+const double LOWEST_EXPECTED_AVAILABLE_AT_END = -0.5;
+const double HIGHEST_EXPECTED_AVAILABLE_AT_END = 0.5;
+
+// These values are the confidence interval expressed as standard deviations.
+// At the minimum confidence level, there is a 25% chance that the true value of
+// the estimate (average cycle time or allocation rate) is not more than
+// MINIMUM_CONFIDENCE standard deviations away from our estimate. Similarly, the
+// MAXIMUM_CONFIDENCE interval here means there is a one in a thousand chance
+// that the true value of our estimate is outside the interval. These are used
+// as bounds on the adjustments applied at the outcome of a GC cycle.
+const double ShenandoahReactiveHeuristics::MINIMUM_CONFIDENCE = 0.319; // 25%
+const double ShenandoahReactiveHeuristics::MAXIMUM_CONFIDENCE = 3.291; // 99.9%
+
+ShenandoahReactiveHeuristics::ShenandoahReactiveHeuristics() :
+  ShenandoahAdaptiveHeuristics(),
+  _allocation_rate(this),
+  _available_at_cycle_start(0),
+  _margin_of_error_sd(ShenandoahReactiveInitialConfidence),
+  _spike_threshold_sd(ShenandoahReactiveInitialSpikeThreshold),
+  _last_trigger(OTHER) { }
+
+ShenandoahReactiveHeuristics::~ShenandoahReactiveHeuristics() {}
+
+void ShenandoahReactiveHeuristics::record_cycle_start() {
+  ShenandoahAdaptiveHeuristics::record_cycle_start();
+  _allocation_rate.allocation_counter_reset();
+  _available_at_cycle_start = ShenandoahHeap::heap()->free_set()->available();
+}
+
+bool ShenandoahReactiveHeuristics::should_start_gc() const {
+  ShenandoahHeap* heap = ShenandoahHeap::heap();
+  size_t capacity = heap->max_capacity();
+  size_t available = heap->free_set()->available();
+  size_t bytes_allocated_since_gc_start = heap->bytes_allocated_since_gc_start();
+
+  // Track allocation rate even if we decide to start a cycle for other reasons.
+  ShenandoahReactiveHeuristics *heuristic = const_cast<ShenandoahReactiveHeuristics *>(this);
```

Comment:
Why this cast is needed? Aren't local fields accessible directly, even without `heuristics->`, or `this->`, for that matter?

Reply:
`should_start_gc` is declared `const`, so it's cast away here. I didn't want to change the signature of `should_start_gc`, but if it were not `const` we could get rid of the cast here.

Reply:
Right. Let's drop the `const` from declarations then!